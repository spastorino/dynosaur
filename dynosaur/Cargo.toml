[package]
name = "dynosaur"
version = "0.2.0"
authors = [
    "Santiago Pastorino <spastorino@gmail.com>",
    "Tyler Mandry <tmandry@google.com>",
    "Niko Matsakis <rust@nikomatsakis.com>",
]
categories = ["asynchronous", "no-std"]
keywords = ["async", "trait", "impl"]
description = "Dynamic dispatch for return position impl traits and async in Rust"
documentation = "https://docs.rs/dynosaur"
license = "MIT OR Apache-2.0"
repository = "https://github.com/spastorino/dynosaur"
edition = "2021"
rust-version = "1.75"

[dependencies]
dynosaur_derive = { version = "0.2", path = "../dynosaur_derive" }

[dev-dependencies]
tokio = { workspace = true }
trait-variant = { workspace = true }
ui_test = "0.28"

[[example]]
name = "next"

[[example]]
name = "trait_variant"

[[test]]
name = "ui"
harness = false
