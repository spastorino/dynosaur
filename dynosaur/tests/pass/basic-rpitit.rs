use dynosaur::dynosaur;

#[dynosaur(DynMyTrait = dyn(box) MyTrait)]
trait MyTrait {
    fn foo(&self) -> impl Send;
}

fn main() {}
