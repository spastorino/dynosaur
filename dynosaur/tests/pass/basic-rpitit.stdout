#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use dynosaur::dynosaur;

trait MyTrait {
    fn foo(&self)
    -> impl Send;
}
mod __dynosaur_macro_dynmytrait {
    use super::*;
    trait ErasedMyTrait {
        fn foo<'life0, 'dynosaur>(&'life0 self)
        -> Box<dyn Send + 'dynosaur>
        where
        'life0: 'dynosaur,
        Self: 'dynosaur;
    }
    impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
        fn foo<'life0, 'dynosaur>(&'life0 self) -> Box<dyn Send + 'dynosaur>
            where 'life0: 'dynosaur, Self: 'dynosaur {
            Box::new(<Self as MyTrait>::foo(self))
        }
    }
    #[repr(transparent)]
    pub struct DynMyTrait<'dynosaur_struct> {
        ptr: dyn ErasedMyTrait + 'dynosaur_struct,
    }
    impl<'dynosaur_struct> MyTrait for DynMyTrait<'dynosaur_struct> {
        fn foo(&self) -> impl Send {
            let ret: Box<dyn Send + '_> = self.ptr.foo();
            let ret: Box<dyn Send + '_> =
                unsafe { ::core::mem::transmute(ret) };
            ret
        }
    }
    impl<'dynosaur_struct> DynMyTrait<'dynosaur_struct> {
        pub fn new_box(value: impl MyTrait + 'dynosaur_struct)
            -> Box<DynMyTrait<'dynosaur_struct>> {
            let value = Box::new(value);
            let value: Box<dyn ErasedMyTrait + 'dynosaur_struct> = value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn new_arc(value: impl MyTrait + 'dynosaur_struct)
            -> std::sync::Arc<DynMyTrait<'dynosaur_struct>> {
            let value = std::sync::Arc::new(value);
            let value: std::sync::Arc<dyn ErasedMyTrait + 'dynosaur_struct> =
                value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn new_rc(value: impl MyTrait + 'dynosaur_struct)
            -> std::rc::Rc<DynMyTrait<'dynosaur_struct>> {
            let value = std::rc::Rc::new(value);
            let value: std::rc::Rc<dyn ErasedMyTrait + 'dynosaur_struct> =
                value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub const fn from_ref(value: &(impl MyTrait + 'dynosaur_struct))
            -> &DynMyTrait<'dynosaur_struct> {
            let value: &(dyn ErasedMyTrait + 'dynosaur_struct) = &*value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub const fn from_mut(value: &mut (impl MyTrait + 'dynosaur_struct))
            -> &mut DynMyTrait<'dynosaur_struct> {
            let value: &mut (dyn ErasedMyTrait + 'dynosaur_struct) =
                &mut *value;
            unsafe { ::core::mem::transmute(value) }
        }
    }
    impl<'dynosaur_struct> MyTrait for &DynMyTrait<'dynosaur_struct> {
        fn foo(&self) -> impl Send {
            <DynMyTrait<'dynosaur_struct> as MyTrait>::foo(self)
        }
    }
    impl<'dynosaur_struct> MyTrait for &mut DynMyTrait<'dynosaur_struct> {
        fn foo(&self) -> impl Send {
            <DynMyTrait<'dynosaur_struct> as MyTrait>::foo(self)
        }
    }
    impl<'dynosaur_struct> MyTrait for Box<DynMyTrait<'dynosaur_struct>> {
        fn foo(&self) -> impl Send {
            <DynMyTrait<'dynosaur_struct> as MyTrait>::foo(self)
        }
    }
}
use __dynosaur_macro_dynmytrait::DynMyTrait;

fn main() {}
